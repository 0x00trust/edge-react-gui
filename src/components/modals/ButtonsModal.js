// @flow

import * as React from 'react'
import { type AirshipBridge } from 'react-native-airship'

import { showError } from '../services/AirshipInstance.js'
import { MainButton } from '../themed/MainButton.js'
import { ModalCloseArrow, ModalMessage, ModalTitle } from '../themed/ModalParts.js'
import { ThemedModal } from '../themed/ThemedModal.js'

type ButtonInfo = {
  label: string,
  type?: 'primary' | 'secondary',

  // The modal will show a spinner as long as this promise is pending.
  // Returning true will dismiss the modal,
  // but returning false will leave the modal up.
  // Although multiple buttons can be spinning at once,
  // a spinning button cannot be clicked again until the promise resolves.
  onPress?: () => Promise<boolean>
}

/**
 * A modal with a title, message, and buttons.
 * This is an alternative to the native `Alert` component.
 *
 * Child components appear between the message and the buttons,
 * but this feature is only meant for inserting extra message elements,
 * like images or custom text formatting.
 *
 * Build a custom modal component if you need form fields, check boxes,
 * or other interactive elements.
 */
export function ButtonsModal<Buttons: { [key: string]: ButtonInfo }>(props: {|
  bridge: AirshipBridge<$Keys<Buttons> | void>,
  title?: string,
  message?: string,
  children?: React.Node,
  buttons: Buttons,
  closeArrow?: boolean,
  disableCancel?: boolean
|}) {
  const { bridge, title, message, children, buttons, closeArrow = false, disableCancel = false } = props

  const handleCancel = disableCancel ? () => {} : () => bridge.resolve(undefined)

  return (
    <ThemedModal bridge={bridge} paddingRem={1} onCancel={handleCancel}>
      {title != null ? <ModalTitle>{title}</ModalTitle> : null}
      {message != null ? <ModalMessage>{message}</ModalMessage> : null}
      {children}
      {Object.keys(buttons).map(key => {
        const { type = 'primary', label, onPress } = buttons[key]

        const handlePress = (): void | Promise<void> => {
          if (onPress == null) return bridge.resolve(key)
          return onPress().then(
            result => {
              if (result) bridge.resolve(key)
            },
            error => showError(error)
          )
        }

        return <MainButton key={key} label={label} marginRem={0.5} type={type} onPress={handlePress} />
      })}
      {closeArrow ? <ModalCloseArrow onPress={handleCancel} /> : null}
    </ThemedModal>
  )
}
