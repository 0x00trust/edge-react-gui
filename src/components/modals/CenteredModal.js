// @flow

import * as React from 'react'
import { Dimensions, View } from 'react-native'
import { type AirshipBridge } from 'react-native-airship'

import { type Theme, cacheStyles, useTheme } from '../services/ThemeContext'
import { ThemedModal } from '../themed/ThemedModal'

type Props = {
  children: React.Node,
  bridge: AirshipBridge<boolean>
}

export function CenteredModal({ children, bridge }: Props) {
  const theme = useTheme()
  const styles = getStyles(theme)

  return (
    <ThemedModal bridge={bridge} position="center" onCancel={() => bridge.resolve(false)}>
      <View style={styles.container}>{children}</View>
    </ThemedModal>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  container: {
    width: Dimensions.get('window').width,
    height: Dimensions.get('window').width
  }
}))
