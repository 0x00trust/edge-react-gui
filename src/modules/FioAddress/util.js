// @flow

import type { EdgeCurrencyConfig, EdgeCurrencyWallet } from 'edge-core-js'

import { FIO_WALLET_TYPE } from '../../constants/WalletAndCurrencyConstants'
import s from '../../locales/strings'
import type { FioConnectionWalletItem, GuiWallet } from '../../types/types'

export const isPubAddressNotConnected = (pubAddress: string | null): boolean => {
  return !pubAddress || pubAddress === '0'
}

export const refreshPubAddressesForFioAddress = async (
  fioAddress: string,
  fioWallet: EdgeCurrencyWallet,
  wallets: EdgeCurrencyWallet[]
): { [fullCurrencyCode: string]: string } => {
  const pubAddresses = {}
  for (const wallet of wallets) {
    const enabledTokens = await wallet.getEnabledTokens()
    if (enabledTokens && enabledTokens.length) {
      for (const enabledToken: string of enabledTokens) {
        try {
          const { public_address } = await fioWallet.otherMethods.fioAction('getPublicAddress', {
            fioAddress,
            tokenCode: enabledToken,
            chainCode: wallet.currencyInfo.currencyCode
          })
          pubAddresses[`${wallet.currencyInfo.currencyCode}:${enabledToken}`] = public_address
        } catch (e) {
          //
          console.log(e.json)
        }
      }
    }
    const fullCurrencyCode = `${wallet.currencyInfo.currencyCode}:${wallet.currencyInfo.currencyCode}`
    if (pubAddresses[fullCurrencyCode] && pubAddresses[fullCurrencyCode] !== '0') continue
    try {
      const { public_address } = await fioWallet.otherMethods.fioAction('getPublicAddress', {
        fioAddress,
        tokenCode: wallet.currencyInfo.currencyCode,
        chainCode: wallet.currencyInfo.currencyCode
      })
      pubAddresses[fullCurrencyCode] = public_address
    } catch (e) {
      //
      console.log(e.json)
    }
  }
  return pubAddresses
}

export const updatePubAddressesForFioAddress = async (
  fioWallet: EdgeCurrencyWallet | null,
  fioAddress: string,
  publicAddresses: { chainCode: string, tokenCode: string, publicAddress: string }[]
) => {
  if (!fioWallet) throw new Error(s.strings.fio_connect_wallets_err)
  let publicAddressesToConnect = []
  const limitPerCall = 5
  for (const { chainCode, tokenCode, publicAddress } of publicAddresses) {
    publicAddressesToConnect.push({
      token_code: tokenCode,
      chain_code: chainCode,
      public_address: publicAddress
    })
    if (publicAddressesToConnect.length === limitPerCall) {
      await addPublicAddresses(fioWallet, fioAddress, publicAddressesToConnect)
      publicAddressesToConnect = []
    }
  }

  if (publicAddressesToConnect.length) {
    await addPublicAddresses(fioWallet, fioAddress, publicAddressesToConnect)
  }
}

export const addPublicAddresses = async (
  fioWallet: EdgeCurrencyWallet,
  fioAddress: string,
  publicAddresses: { token_code: string, chain_code: string, public_address: string }[]
) => {
  let maxFee: number
  try {
    const { fee } = await fioWallet.otherMethods.fioAction('getFeeForAddPublicAddress', {
      fioAddress
    })
    maxFee = fee
  } catch (e) {
    throw new Error(s.strings.fio_get_fee_err_msg)
  }
  try {
    await fioWallet.otherMethods.fioAction('addPublicAddresses', {
      fioAddress,
      publicAddresses,
      maxFee
    })
  } catch (e) {
    throw new Error(s.strings.fio_connect_wallets_err)
  }
}

export const findWalletByFioAddress = async (fioWallets: EdgeCurrencyWallet[], fioAddress: string): Promise<EdgeCurrencyWallet | null> => {
  if (fioWallets) {
    for (const wallet: EdgeCurrencyWallet of fioWallets) {
      const fioAddresses: string[] = await wallet.otherMethods.getFioAddressNames()
      for (const address of fioAddresses) {
        if (address.toLowerCase() === fioAddress.toLowerCase()) {
          return wallet
        }
      }
    }
  }

  return null
}

export const makeNotConnectedWallets = (
  wallets: { [walletId: string]: GuiWallet },
  connectedPubAddresses: { [fullCurrencyCode: string]: string }
): { [key: string]: FioConnectionWalletItem } => {
  const notConnectedWallets = {}
  for (const walletKey: string in wallets) {
    if (wallets[walletKey].type === FIO_WALLET_TYPE) continue
    const publicAddress = wallets[walletKey].receiveAddress.publicAddress
    if (!publicAddress) continue
    const fullCurrencyCode = `${wallets[walletKey].currencyCode}:${wallets[walletKey].currencyCode}`
    if (isPubAddressNotConnected(connectedPubAddresses[fullCurrencyCode])) {
      notConnectedWallets[`${wallets[walletKey].id}-${wallets[walletKey].currencyCode}`] = {
        key: `${wallets[walletKey].id}-${wallets[walletKey].currencyCode}`,
        id: wallets[walletKey].id,
        publicAddress,
        symbolImage: wallets[walletKey].symbolImage,
        name: wallets[walletKey].name,
        currencyCode: wallets[walletKey].currencyCode,
        chainCode: wallets[walletKey].currencyCode,
        fullCurrencyCode
      }
    }
    if (wallets[walletKey].enabledTokens && wallets[walletKey].enabledTokens.length) {
      for (const enabledToken: string of wallets[walletKey].enabledTokens) {
        const tokenData = wallets[walletKey].metaTokens.find(metaToken => metaToken.currencyCode === enabledToken)
        if (!tokenData) continue
        const fullCurrencyCode = `${wallets[walletKey].currencyCode}:${tokenData.currencyCode}`
        if (isPubAddressNotConnected(connectedPubAddresses[fullCurrencyCode])) {
          notConnectedWallets[`${wallets[walletKey].id}-${tokenData.currencyCode}`] = {
            key: `${wallets[walletKey].id}-${tokenData.currencyCode}`,
            id: wallets[walletKey].id,
            publicAddress,
            symbolImage: tokenData.symbolImage,
            name: wallets[walletKey].name,
            currencyCode: tokenData.currencyCode,
            chainCode: wallets[walletKey].currencyCode,
            fullCurrencyCode
          }
        }
      }
    }
  }

  return notConnectedWallets
}

export const makeConnectedWallets = (
  wallets: { [walletId: string]: GuiWallet },
  connectedPubAddresses: { [fullCurrencyCode: string]: string }
): { [key: string]: FioConnectionWalletItem } => {
  const connectedWallets = {}
  for (const walletKey: string in wallets) {
    if (wallets[walletKey].type === FIO_WALLET_TYPE) continue
    const publicAddress = wallets[walletKey].receiveAddress.publicAddress
    if (!publicAddress) continue
    const fullCurrencyCode = `${wallets[walletKey].currencyCode}:${wallets[walletKey].currencyCode}`
    if (publicAddress === connectedPubAddresses[fullCurrencyCode]) {
      connectedWallets[`${wallets[walletKey].id}-${wallets[walletKey].currencyCode}`] = {
        key: `${wallets[walletKey].id}-${wallets[walletKey].currencyCode}`,
        id: wallets[walletKey].id,
        publicAddress,
        symbolImage: wallets[walletKey].symbolImage,
        name: wallets[walletKey].name,
        currencyCode: wallets[walletKey].currencyCode,
        chainCode: wallets[walletKey].currencyCode,
        fullCurrencyCode
      }
    }
    if (wallets[walletKey].enabledTokens && wallets[walletKey].enabledTokens.length) {
      for (const enabledToken: string of wallets[walletKey].enabledTokens) {
        const tokenData = wallets[walletKey].metaTokens.find(metaToken => metaToken.currencyCode === enabledToken)
        if (!tokenData) continue
        const fullCurrencyCode = `${wallets[walletKey].currencyCode}:${tokenData.currencyCode}`
        if (publicAddress === connectedPubAddresses[fullCurrencyCode]) {
          connectedWallets[`${wallets[walletKey].id}-${tokenData.currencyCode}`] = {
            key: `${wallets[walletKey].id}-${tokenData.currencyCode}`,
            id: wallets[walletKey].id,
            publicAddress,
            symbolImage: tokenData.symbolImage,
            name: wallets[walletKey].name,
            currencyCode: tokenData.currencyCode,
            chainCode: wallets[walletKey].currencyCode,
            fullCurrencyCode
          }
        }
      }
    }
  }

  return connectedWallets
}

export const checkPubAddress = async (fioPlugin: EdgeCurrencyConfig, fioAddress: string, chainCode: string, tokenCode: string): Promise<string> => {
  const isFioAddress = await fioPlugin.otherMethods.isFioAddressValid(fioAddress)
  try {
    if (isFioAddress) {
      const { public_address: publicAddress } = await fioPlugin.otherMethods.getConnectedPublicAddress(fioAddress.toLowerCase(), chainCode, tokenCode)
      if (publicAddress && publicAddress.length > 1) {
        return publicAddress
      }
    }
  } catch (e) {
    throw new Error(s.strings.err_no_address_title)
  }

  return ''
}
