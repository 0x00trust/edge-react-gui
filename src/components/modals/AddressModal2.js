// @flow

import { FormField, TertiaryButton } from 'edge-components'
import type { EdgeAccount, EdgeCurrencyConfig, EdgeCurrencyWallet } from 'edge-core-js'
import React, { Component, Fragment } from 'react'
import { Clipboard, FlatList, Image, InputAccessoryView, TouchableOpacity, TouchableWithoutFeedback, View } from 'react-native'
import FontAwesomeIcon from 'react-native-vector-icons/FontAwesome'
import { sprintf } from 'sprintf-js'

import ENS_LOGO from '../../assets/images/ens_logo.png'
import FIO_LOGO from '../../assets/images/fio_logo.png'
import s from '../../locales/strings.js'
import { checkPubAddress, getFioAddressCache } from '../../modules/FioAddress/util.js'
import Text from '../../modules/UI/components/FormattedText/index'
import styles, { addressInputStyles, iconStyles } from '../../styles/components/ScanAddressModal.js'
import type { FlatListItem } from '../../types/types.js'
import ResolutionError, { ResolutionErrorCode } from '../common/ResolutionError.js'
import { type AirshipBridge, AirshipModal, dayText, IconCircle } from './modalParts.js'

const MODAL_ICON = 'address-book-o'
const inputAccessoryViewID: string = 'inputAccessoryViewID'

export type AddressModalOpts = {
  walletId: string,
  coreWallet: EdgeCurrencyWallet,
  fioPlugin: EdgeCurrencyConfig,
  currencyCode: string
}

type Props = {
  bridge: AirshipBridge<string | null>,
  coreWallet: EdgeCurrencyWallet,
  fioPlugin: EdgeCurrencyConfig,
  currencyCode: string,
  account: EdgeAccount
}

type State = {
  clipboard: string,
  uri: string,
  statusLabel: string,
  fieldError: string,
  cryptoAddress?: string,
  fioAddresses: string[]
}

export class AddressModal2 extends Component<Props, State> {
  fioCheckQueue: number = 0

  constructor (props: Props) {
    super(props)
    this.fioCheckQueue = 0
    this.state = {
      clipboard: '',
      uri: '',
      statusLabel: s.strings.fragment_send_address,
      cryptoAddress: undefined,
      fieldError: '',
      fioAddresses: []
    }
  }

  componentDidMount () {
    this._setClipboard(this.props)
    this.setFioAddresses()
  }

  _setClipboard = async props => {
    const coreWallet = props.coreWallet

    try {
      const uri = await Clipboard.getString()

      // Will throw in case uri is invalid
      await coreWallet.parseUri(uri)

      this.setState({
        clipboard: uri
      })
    } catch (e) {
      // Failure is acceptable
    }
  }

  setFioAddresses = async () => {
    const fioAddressesObject = await getFioAddressCache(this.props.account)
    const fioAddresses = fioAddressesObject.addresses ? Object.keys(fioAddressesObject.addresses) : []
    this.setState({ fioAddresses: fioAddresses.sort() })
  }

  setStatusLabel = (status: string) => {
    this.setState({ statusLabel: status })
  }

  setCryptoAddress = (address?: string) => {
    this.setState({ cryptoAddress: address })
  }

  onChangeTextDelayed = (domain: string) => {
    const { currencyCode } = this.props
    if (this.checkIfDomain(domain)) {
      this.resolveAddress(domain, currencyCode)
    }
    this.checkIfFioAddress(domain)
    this.updateUri(domain)
  }

  checkIfDomain = (domain: string): boolean => {
    return this.checkIfUnstoppableDomain(domain) || this.checkIfEnsDomain(domain)
  }

  checkIfUnstoppableDomain = (domain: string): boolean => {
    return domain.endsWith('.zil') || domain.endsWith('.crypto')
  }

  checkIfEnsDomain = (domain: string): boolean => {
    return domain.endsWith('.eth') || domain.endsWith('.luxe') || domain.endsWith('.kred') || domain.endsWith('.xyz')
  }

  fetchDomain = async (domain: string, currencyTicker: string): Promise<string> => {
    domain = domain.trim().toLowerCase()
    if (!this.checkIfDomain(domain)) {
      throw new ResolutionError(ResolutionErrorCode.UnsupportedDomain, { domain })
    }
    const baseurl = `https://unstoppabledomains.com/api/v1`
    const url = this.checkIfEnsDomain(domain) ? `${baseurl}/${domain}/${currencyTicker}` : `${baseurl}/${domain}`
    const response = await global.fetch(url).then(res => res.json())
    const { addresses, meta } = response
    if (!meta || !meta.owner) {
      throw new ResolutionError(ResolutionErrorCode.UnregisteredDomain, { domain })
    }
    const ticker = currencyTicker.toUpperCase()
    if (!addresses || !addresses[ticker]) {
      throw new ResolutionError(ResolutionErrorCode.UnspecifiedCurrency, { domain, currencyTicker })
    }
    return addresses[ticker]
  }

  resolveAddress = async (domain: string, currencyTicker: string) => {
    if (!domain) return
    try {
      this.setStatusLabel(s.strings.resolving)
      const addr = await this.fetchDomain(domain, currencyTicker)
      this.setStatusLabel(addr)
      this.setCryptoAddress(addr)
    } catch (err) {
      if (err instanceof ResolutionError) {
        const message = sprintf(s.strings[err.code], domain, currencyTicker)
        if (domain === '') this.setStatusLabel(s.strings.fragment_send_address)
        else {
          this.setStatusLabel(message)
          this.setCryptoAddress(undefined)
        }
      }
    }
  }

  checkFioPubAddressQueue (uri: string) {
    this.setStatusLabel(s.strings.resolving)
    this.fioCheckQueue++
    setTimeout(async () => {
      // do not check if user continue typing fio address
      if (this.fioCheckQueue > 1) {
        return --this.fioCheckQueue
      }
      this.fioCheckQueue = 0
      try {
        const { currencyCode, coreWallet, fioPlugin } = this.props
        const publicAddress = await checkPubAddress(fioPlugin, uri.toLowerCase(), coreWallet.currencyInfo.currencyCode, currencyCode)
        this.setStatusLabel(s.strings.fragment_send_address)
        if (!publicAddress) {
          return this.setState({ fieldError: s.strings.err_no_address_title })
        }
      } catch (e) {
        this.setStatusLabel(s.strings.fragment_send_address)
        return this.setState({ fieldError: e.message })
      }
    }, 1000)
  }

  async checkIfFioAddress (uri: string) {
    this.setState({ fieldError: '' })

    if (await this.isFioAddressValid(uri)) {
      this.checkFioPubAddressQueue(uri)
    }
  }

  isFioAddressValid = (fioAddress: string) => {
    const { fioPlugin } = this.props
    return fioPlugin.otherMethods.isFioAddressValid(fioAddress)
  }

  updateUri = (uri: string) => {
    this.setState({
      uri
    })
  }

  onPasteFromClipboard = () => {
    const { clipboard } = this.state
    this.setState({ uri: clipboard }, async () => {
      if (await this.isFioAddressValid(clipboard)) {
        await this.checkIfFioAddress(clipboard)
      }
      this.handleSubmit()
    })
  }

  onPressFioAddress = (address: string) => {
    this.setState({ uri: address }, async () => {
      if (await this.isFioAddressValid(address)) {
        await this.checkIfFioAddress(address)
      }
      this.handleSubmit()
    })
  }

  renderFioAddressRow = ({ item }: FlatListItem<string>) => {
    let addressType
    if (this.checkIfDomain(item)) {
      addressType = ENS_LOGO
    } else if (item.includes('@')) {
      addressType = FIO_LOGO
    } else {
      return null
    }
    return (
      <TouchableWithoutFeedback onPress={() => this.onPressFioAddress(item)}>
        <View style={styles.tileContainer}>
          <Image source={addressType} style={styles.fioAddressAvatarContainer} resizeMode={'cover'} />
          <Text style={styles.fioAddressText}>{item}</Text>
        </View>
      </TouchableWithoutFeedback>
    )
  }

  handleSubmit = () => {
    const { uri, cryptoAddress, fieldError } = this.state
    const submitData = cryptoAddress || uri
    if (fieldError) return
    this.props.bridge.resolve(submitData)
  }
  handleClose = () => this.props.bridge.resolve(null)
  keyExtractor = (item: string, index: number) => index.toString()
  render () {
    const copyMessage = this.state.clipboard ? sprintf(s.strings.string_paste_address, this.state.clipboard) : null
    const { uri, statusLabel, fieldError } = this.state
    return (
      <AirshipModal bridge={this.props.bridge} onCancel={this.handleClose}>
        {gap => (
          <Fragment>
            <IconCircle>
              <FontAwesomeIcon name={MODAL_ICON} size={iconStyles.size} color={iconStyles.color} />
            </IconCircle>
            <View style={styles.container}>
              <View style={styles.tileContainerHeader}>
                <Text style={dayText('title')}>{s.strings.scan_address_modal_header}</Text>
                <Text style={dayText('title')}>{s.strings.scan_address_modal_sub_header}</Text>
              </View>
              {copyMessage && (
                <View style={styles.tileContainerButtons}>
                  <TertiaryButton ellipsizeMode={'middle'} onPress={this.onPasteFromClipboard} numberOfLines={1} style={styles.addressModalButton}>
                    <TertiaryButton.Text>{copyMessage}</TertiaryButton.Text>
                  </TertiaryButton>
                </View>
              )}
              <View style={styles.tileContainerInput}>
                <InputAccessoryView nativeID={inputAccessoryViewID}>
                  <View style={styles.accessoryView}>
                    <TouchableOpacity style={styles.accessoryBtn} onPress={this.handleClose}>
                      <Text style={styles.accessoryText}>{s.strings.string_cancel_cap}</Text>
                    </TouchableOpacity>
                  </View>
                </InputAccessoryView>
                <FormField
                  autoFocus
                  blurOnSubmit
                  returnKeyType="done"
                  autoCapitalize="none"
                  autoCorrect={false}
                  style={addressInputStyles}
                  value={uri}
                  onChangeText={this.onChangeTextDelayed}
                  error={fieldError}
                  label={statusLabel}
                  onSubmitEditing={this.handleSubmit}
                  inputAccessoryViewID={inputAccessoryViewID}
                />
              </View>
              <FlatList
                style={{ flex: 1, marginBottom: -gap.bottom }}
                contentContainerStyle={{ paddingBottom: gap.bottom }}
                data={this.state.fioAddresses}
                initialNumToRender={24}
                keyboardShouldPersistTaps="handled"
                keyExtractor={this.keyExtractor}
                renderItem={this.renderFioAddressRow}
              />
            </View>
          </Fragment>
        )}
      </AirshipModal>
    )
  }
}
