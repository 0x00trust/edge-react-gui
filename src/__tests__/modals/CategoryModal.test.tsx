import { describe, expect, it } from '@jest/globals'
import * as React from 'react'
import { createRenderer } from 'react-test-renderer/shallow'

import { CategoryModal } from '../../components/modals/CategoryModal'
import s from '../../locales/strings'
import { defaultCategories } from '../../util/categories'
import { fakeAirshipBridge } from '../../util/fake/fakeAirshipBridge'

describe('CategoryModal', () => {
  const categories = {
    exchange: {
      syntax: s.strings.fragment_transaction_exchange,
      key: 'exchange'
    },
    expense: {
      syntax: s.strings.fragment_transaction_expense,
      key: 'expense'
    },
    transfer: {
      syntax: s.strings.fragment_transaction_transfer,
      key: 'transfer'
    },
    income: {
      syntax: s.strings.fragment_transaction_income,
      key: 'income'
    }
  }

  it('should render with an empty subcategory', () => {
    const renderer = createRenderer()

    const actual = renderer.render(
      <CategoryModal
        bridge={fakeAirshipBridge}
        categories={categories}
        subCategories={defaultCategories.slice(0, 4)}
        category="exchange"
        subCategory=""
        setNewSubcategory={(input: string, subCategories: string[]) => undefined}
      />
    )

    expect(actual).toMatchSnapshot()
  })

  it('should render with a subcategory', () => {
    const renderer = createRenderer()

    const actual = renderer.render(
      <CategoryModal
        bridge={fakeAirshipBridge}
        categories={categories}
        subCategories={defaultCategories.slice(0, 4)}
        category="income"
        subCategory="Paycheck"
        setNewSubcategory={(input: string, subCategories: string[]) => undefined}
      />
    )

    expect(actual).toMatchSnapshot()
  })
})
