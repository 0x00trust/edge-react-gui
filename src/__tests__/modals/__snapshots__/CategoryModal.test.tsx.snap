// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryModal should render with a subcategory 1`] = `
<AirshipModal
  bridge={
    Object {
      "on": [Function],
      "onResult": [Function],
      "reject": [Function],
      "remove": [Function],
      "resolve": [Function],
    }
  }
  onCancel={[Function]}
>
  <TouchableWithoutFeedback
    onPress={[Function]}
  >
    <View
      style={
        Object {
          "flex": 1,
          "padding": 18,
        }
      }
    >
      <FormattedText
        style={
          Object {
            "alignSelf": "center",
            "fontSize": 27,
            "marginBottom": 22,
          }
        }
      >
        Category
      </FormattedText>
      <View
        style={
          Object {
            "marginBottom": 18,
          }
        }
      >
        <FormattedText
          style={
            Object {
              "color": "#0E4B75",
              "fontSize": 16,
              "marginBottom": 7,
            }
          }
        >
          Choose a Category:
        </FormattedText>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <TouchableWithoutFeedback
            onPress={[Function]}
          >
            <View
              style={
                Object {
                  "borderColor": "#A4C7DF",
                  "borderRadius": 3,
                  "borderWidth": 1,
                  "marginRight": 13,
                  "paddingHorizontal": 11,
                  "paddingVertical": 4,
                }
              }
            >
              <FormattedText
                style={
                  Object {
                    "color": "#0E4B75",
                    "fontSize": 20,
                  }
                }
              >
                Exchange
              </FormattedText>
            </View>
          </TouchableWithoutFeedback>
          <TouchableWithoutFeedback
            onPress={[Function]}
          >
            <View
              style={
                Object {
                  "borderColor": "#A4C7DF",
                  "borderRadius": 3,
                  "borderWidth": 1,
                  "marginRight": 13,
                  "paddingHorizontal": 11,
                  "paddingVertical": 4,
                }
              }
            >
              <FormattedText
                style={
                  Object {
                    "color": "#0E4B75",
                    "fontSize": 20,
                  }
                }
              >
                Expense
              </FormattedText>
            </View>
          </TouchableWithoutFeedback>
          <TouchableWithoutFeedback
            onPress={[Function]}
          >
            <View
              style={
                Object {
                  "borderColor": "#A4C7DF",
                  "borderRadius": 3,
                  "borderWidth": 1,
                  "marginRight": 13,
                  "paddingHorizontal": 11,
                  "paddingVertical": 4,
                }
              }
            >
              <FormattedText
                style={
                  Object {
                    "color": "#0E4B75",
                    "fontSize": 20,
                  }
                }
              >
                Transfer
              </FormattedText>
            </View>
          </TouchableWithoutFeedback>
          <TouchableWithoutFeedback
            onPress={[Function]}
          >
            <View
              style={
                Object {
                  "backgroundColor": "#A4C7DF",
                  "borderColor": "#A4C7DF",
                  "borderRadius": 3,
                  "borderWidth": 1,
                  "marginRight": 13,
                  "paddingHorizontal": 11,
                  "paddingVertical": 4,
                }
              }
            >
              <FormattedText
                style={
                  Object {
                    "color": "#0E4B75",
                    "fontSize": 20,
                  }
                }
              >
                Income
              </FormattedText>
            </View>
          </TouchableWithoutFeedback>
        </View>
      </View>
      <View
        style={
          Object {
            "marginTop": 18,
          }
        }
      >
        <FormField
          autoCapitalize="none"
          autoCorrect={false}
          autoFocus={true}
          baseColor="#0D2145"
          containerStyle={
            Object {
              "height": 77,
              "position": "relative",
              "width": "100%",
            }
          }
          errorColor="#E85466"
          fontSize={20}
          keyboardType="default"
          label="Choose a sub-category"
          labelFontSize={15}
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          returnKeyType="done"
          textColor="#25292C"
          tintColor="#0E4B75"
          titleTextStyle={Object {}}
          value="Paycheck"
        />
      </View>
      <SubCategorySelect
        bottomGap={0}
        categories={
          Array [
            "income",
            "exchange",
            "expense",
            "transfer",
          ]
        }
        enteredSubcategory="Paycheck"
        onPressFxn={[Function]}
        subcategoriesList={
          Array [
            "Exchange:Buy Bitcoin",
            "Exchange:Sell Bitcoin",
            "Expense:Air Travel",
            "Expense:Alcohol & Bars",
          ]
        }
      />
    </View>
  </TouchableWithoutFeedback>
</AirshipModal>
`;

exports[`CategoryModal should render with an empty subcategory 1`] = `
<AirshipModal
  bridge={
    Object {
      "on": [Function],
      "onResult": [Function],
      "reject": [Function],
      "remove": [Function],
      "resolve": [Function],
    }
  }
  onCancel={[Function]}
>
  <TouchableWithoutFeedback
    onPress={[Function]}
  >
    <View
      style={
        Object {
          "flex": 1,
          "padding": 18,
        }
      }
    >
      <FormattedText
        style={
          Object {
            "alignSelf": "center",
            "fontSize": 27,
            "marginBottom": 22,
          }
        }
      >
        Category
      </FormattedText>
      <View
        style={
          Object {
            "marginBottom": 18,
          }
        }
      >
        <FormattedText
          style={
            Object {
              "color": "#0E4B75",
              "fontSize": 16,
              "marginBottom": 7,
            }
          }
        >
          Choose a Category:
        </FormattedText>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <TouchableWithoutFeedback
            onPress={[Function]}
          >
            <View
              style={
                Object {
                  "backgroundColor": "#A4C7DF",
                  "borderColor": "#A4C7DF",
                  "borderRadius": 3,
                  "borderWidth": 1,
                  "marginRight": 13,
                  "paddingHorizontal": 11,
                  "paddingVertical": 4,
                }
              }
            >
              <FormattedText
                style={
                  Object {
                    "color": "#0E4B75",
                    "fontSize": 20,
                  }
                }
              >
                Exchange
              </FormattedText>
            </View>
          </TouchableWithoutFeedback>
          <TouchableWithoutFeedback
            onPress={[Function]}
          >
            <View
              style={
                Object {
                  "borderColor": "#A4C7DF",
                  "borderRadius": 3,
                  "borderWidth": 1,
                  "marginRight": 13,
                  "paddingHorizontal": 11,
                  "paddingVertical": 4,
                }
              }
            >
              <FormattedText
                style={
                  Object {
                    "color": "#0E4B75",
                    "fontSize": 20,
                  }
                }
              >
                Expense
              </FormattedText>
            </View>
          </TouchableWithoutFeedback>
          <TouchableWithoutFeedback
            onPress={[Function]}
          >
            <View
              style={
                Object {
                  "borderColor": "#A4C7DF",
                  "borderRadius": 3,
                  "borderWidth": 1,
                  "marginRight": 13,
                  "paddingHorizontal": 11,
                  "paddingVertical": 4,
                }
              }
            >
              <FormattedText
                style={
                  Object {
                    "color": "#0E4B75",
                    "fontSize": 20,
                  }
                }
              >
                Transfer
              </FormattedText>
            </View>
          </TouchableWithoutFeedback>
          <TouchableWithoutFeedback
            onPress={[Function]}
          >
            <View
              style={
                Object {
                  "borderColor": "#A4C7DF",
                  "borderRadius": 3,
                  "borderWidth": 1,
                  "marginRight": 13,
                  "paddingHorizontal": 11,
                  "paddingVertical": 4,
                }
              }
            >
              <FormattedText
                style={
                  Object {
                    "color": "#0E4B75",
                    "fontSize": 20,
                  }
                }
              >
                Income
              </FormattedText>
            </View>
          </TouchableWithoutFeedback>
        </View>
      </View>
      <View
        style={
          Object {
            "marginTop": 18,
          }
        }
      >
        <FormField
          autoCapitalize="none"
          autoCorrect={false}
          autoFocus={true}
          baseColor="#0D2145"
          containerStyle={
            Object {
              "height": 77,
              "position": "relative",
              "width": "100%",
            }
          }
          errorColor="#E85466"
          fontSize={20}
          keyboardType="default"
          label="Choose a sub-category"
          labelFontSize={15}
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          returnKeyType="done"
          textColor="#25292C"
          tintColor="#0E4B75"
          titleTextStyle={Object {}}
          value=""
        />
      </View>
      <SubCategorySelect
        bottomGap={0}
        categories={
          Array [
            "exchange",
            "expense",
            "transfer",
            "income",
          ]
        }
        enteredSubcategory=""
        onPressFxn={[Function]}
        subcategoriesList={
          Array [
            "Exchange:Buy Bitcoin",
            "Exchange:Sell Bitcoin",
            "Expense:Air Travel",
            "Expense:Alcohol & Bars",
          ]
        }
      />
    </View>
  </TouchableWithoutFeedback>
</AirshipModal>
`;
